### PR TITLE
cleanly partition pls.h as own dependency

### PR DESCRIPTION
**Pull request by pearsonca**
This change partitions the pls.h functionality as its own dependency, i.e. without any other abcsmc library dependencies.

This has the knock-on effect of moving some definitions from AbcUtil.h to pls.h, namely the linear algebra types, some functions which are used (within AbcSmc, at least - not sure about simulators) exclusively in pls functions, and one function which was used elsewhere (`ordered`).

I think there's an argument to be made for a "PLS" namespace to capture the types, but didn't do that here. I also think there's some argument for moving more of AbcUtil functions to PLS header. Those should be straightforward changes.

~~Once we've resolved those considerations, the next level step is of course to spin PLS.h off entirely into it's own repository, and then refactor this to depend on that as a library.~~

### CODE REVIEW

**Comment by pearsonca** [comment]
Also tempted to incorporate this for signum operations: https://stackoverflow.com/a/4609795/167973

**Comment by pearsonca** [comment]
n.b. there is a [PLS](https://github.com/tjhladish/PLS) library already, and this change moves in the direction of having that as a well-defined dependency (for simplicity, perhaps incorporated as a submodule?)

File: pls.h [hunk 1]
```diff
@@ -38,8 +52,134 @@ typedef enum { LOO, NEW_DATA } VALIDATION_METHOD;
...
+    return abs(es.eigenvalues()[idx].real());
+};
+
+
+Colc dominant_eigenvector( EigenSolver<Mat2D> es ){
+    const size_t idx = find_dominant_ev(es);
+    return es.eigenvectors().col(idx);
+}
+
+#include <algorithm> // sort
```
Comment (tjhladish) [inline pls.h:85]: Pls move includes to top

File: pls.h [hunk 2]
```diff
@@ -38,8 +52,134 @@ typedef enum { LOO, NEW_DATA } VALIDATION_METHOD;
...
+#include <numeric> // iota
+
+// tag/index sort, from https://stackoverflow.com/a/37732329/167973
+template<typename T>
+std::vector<std::size_t> ordered(const T& v)
```
Comment (tjhladish) [inline pls.h:90]: Standardize whitespace usage (not just here): `ret_type func_name (arg_type arg) {`

File: AbcSim.h [hunk 3]
```diff
@@ -40,16 +40,16 @@ namespace ABC {
...
     ) const = 0;
 };
 
 // an AbcSimFun which throws an error if used. This is intended to be used as a default, so that if *not* replaced, the error
 // will be thrown when the simulator is used.
 struct AbcSimUnset : AbcSimFun {
-    vector<ABC::float_type> operator()(
-      vector<ABC::float_type> /*pars*/, const unsigned long int /*seed*/, const unsigned long int /*serial*/, const ABC::MPI_par* /*_mp*/
+    vector<float_type> operator()(
+      vector<float_type> pars, const unsigned long int seed, const unsigned long int serial, const ABC::MPI_par* _mp
```
Comment (tjhladish) [inline AbcSim.h:52]: @pearsonca I think we still want the args commented:
```suggestion
      vector<float_type> /*pars*/, const unsigned long int /*seed*/, const unsigned long int /*serial*/, const ABC::MPI_par* /*_mp*/
```

File: AbcSim.h [hunk 4]
```diff
@@ -98,11 +98,11 @@ struct AbcExec : AbcSimFun {
     const string command;
     AbcExec(string _command) : command(_command) { }
 
-    vector<ABC::float_type> operator()(
-      vector<ABC::float_type> pars, const unsigned long int /*seed*/, const unsigned long int /*serial*/, const ABC::MPI_par* /*_mp*/
+    vector<float_type> operator()(
+      vector<float_type> pars, const unsigned long int seed, const unsigned long int serial, const ABC::MPI_par* _mp
```
Comment (tjhladish) [inline AbcSim.h:102]:
```suggestion
      vector<float_type> pars, const unsigned long int /*seed*/, const unsigned long int /*serial*/, const ABC::MPI_par* /*_mp*/
```

**Review by tjhladish — CHANGES_REQUESTED** [review]
See in-line suggested changes

**Pull request merged**